### PR TITLE
fix: corrected inflate and deflate logic for xml, and corresponding tests with some refactoring for SAML sessions

### PR DESCRIPTION
**Pull request by stebenz**
Corrected the inflate and deflate logic for the handling of requests and responses in REDIRECT-binding, as they were not correctly handled. Refactoring for SAML sessions on Zitadel side, so that responses, timezone and expiration times of messages can be shared.

### Definition of Ready

- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [x] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] My code has no repetitions
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [x] Functionality of the acceptance criteria is checked manually on the dev system.

### CODE REVIEW

File: pkg/provider/xml/xml.go [hunk 1]
```diff
@@ -148,50 +118,50 @@ func DecodeAttributeQuery(request string) (*samlp.AttributeQueryType, error) {
...
 			return nil, err
 		}
-	default:
-		return nil, fmt.Errorf("unknown encoding")
+		r.Close()
+		return out, nil
 	}
-
-	return req, nil
+	return nil, fmt.Errorf("unknown encoding")
```
Comment (livio-a) [inline pkg/provider/xml/xml.go:166]: I would change it to:
```go
switch encoding {
	case "":
		return data, nil
	case EncodingDeflate:
		r := flate.NewReader(bytes.NewBuffer(data))
		defer r.Close()
		return io.ReadAll(r)
	default:
		return nil, ErrUnknownEncoding
}
```

File: pkg/provider/response.go [hunk 2]
```diff
@@ -168,10 +133,11 @@ func (r *Response) makeFailedResponse(
 func (r *Response) makeSuccessfulResponse(
 	attributes *Attributes,
 	timeFormat string,
+	expiration time.Duration,
 ) *samlp.ResponseType {
 	now := time.Now().UTC()
 	nowStr := now.Format(timeFormat)
-	fiveFromNowStr := now.Add(5 * time.Minute).Format(timeFormat)
+	fiveFromNowStr := now.Add(expiration).Format(timeFormat)
```
Comment (livio-a) [inline pkg/provider/response.go:140]: fiveFromNowStr -> validity?

File: pkg/provider/response.go [hunk 3]
```diff
@@ -12,17 +14,17 @@ import (
...
+	statusCodeSuccess                = errors.New("urn:oasis:names:tc:SAML:2.0:status:Success")
+	StatusCodeVersionMissmatch       = errors.New("urn:oasis:names:tc:SAML:2.0:status:VersionMismatch")
+	StatusCodeAuthNFailed            = errors.New("urn:oasis:names:tc:SAML:2.0:status:AuthnFailed")
+	StatusCodeInvalidAttrNameOrValue = errors.New("urn:oasis:names:tc:SAML:2.0:status:InvalidAttrNameOrValue")
+	StatusCodeInvalidNameIDPolicy    = errors.New("urn:oasis:names:tc:SAML:2.0:status:InvalidNameIDPolicy")
+	StatusCodeRequestDenied          = errors.New("urn:oasis:names:tc:SAML:2.0:status:RequestDenied")
+	StatusCodeRequestUnsupported     = errors.New("urn:oasis:names:tc:SAML:2.0:status:RequestUnsupported")
+	StatusCodeUnsupportedBinding     = errors.New("urn:oasis:names:tc:SAML:2.0:status:UnsupportedBinding")
+	StatusCodeResponder              = errors.New("urn:oasis:names:tc:SAML:2.0:status:Responder")
+	StatusCodePartialLogout          = errors.New("urn:oasis:names:tc:SAML:2.0:status:PartialLogout")
```
Comment (livio-a) [inline pkg/provider/response.go:27]: why is this (breaking) change needed? they're anyway just used as error.Error()

Comment (livio-a) [reply pkg/provider/response.go:27]: also making success an error and private?

File: pkg/provider/response_test.go [hunk 4]
```diff
@@ -94,7 +94,7 @@ func TestResponse_sendBackResponse(t *testing.T) {
...
+				body: []byte("<a href=\"/acs?SAMLResponse=nJLBSgMxEIbvPkWY%2B3bTniQkKcVSWNCLrT14C7ujBHYnNTMr9e2lKduCB7FCLgnz5Z98E7s8Dr36xMwxkYP5TINCalMX6d3By25T3cPS39ln5EMiRnUcemIHYyaTAkc2FAZkI63Zrp4ezWKmzSEnSW3qQTVrB7ED1dDE75KDjB8jsoDaT7GLU2zDPGJDLIHEwUJrXel5pec7rU1Zr6DWyBIpSKFCy%2BBtwfJf%2BgrMmE8oqE3KQ5Dfy08nsaveSqlBkihf4GOJs%2FU51tutBBn5Ji0T9JC6G33uQz%2BiAy44eFtfL7psvF1N77xRymUc58H9nMg%2FXfurq%2FrSmbf19CP8dwAAAP%2F%2F&amp;RelayState=relayState&amp;Signature=sig&amp;SigAlg=alg\">Found</a>.\n\n"),
```
Comment (livio-a) [inline pkg/provider/response_test.go:97]: i guess these changed because of the higher compression level?

**Review by livio-a — CHANGES_REQUESTED** [review]

File: pkg/provider/xml/xml.go [hunk 5]
```diff
@@ -86,53 +81,28 @@ func Write(w http.ResponseWriter, body []byte) error {
...
-	default:
-		reader := flate.NewReader(bytes.NewReader(reqBytes))
-		decoder := xml.NewDecoder(reader)
-		if err = decoder.Decode(req); err != nil {
-			if err := xml.Unmarshal(reqBytes, req); err != nil {
-				return nil, fmt.Errorf("failed to unmarshal: %w", err)
-			}
-		}
+	if err := xml.Unmarshal(data, req); err != nil {
+		fmt.Println(err.Error())
```
[the remark this answers is not in the document]
Comment (stebenz) [reply pkg/provider/xml/xml.go]: removed

File: pkg/provider/xml/xml.go [hunk 6]
```diff
@@ -86,53 +81,28 @@ func Write(w http.ResponseWriter, body []byte) error {
...
-		}
-	default:
-		reader := flate.NewReader(bytes.NewReader(reqBytes))
-		decoder := xml.NewDecoder(reader)
-		if err = decoder.Decode(req); err != nil {
-			if err := xml.Unmarshal(reqBytes, req); err != nil {
-				return nil, fmt.Errorf("failed to unmarshal: %w", err)
-			}
-		}
+	if err := xml.Unmarshal(data, req); err != nil {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply pkg/provider/xml/xml.go:89]: Left out for now, can be replaced and added in the future if necessary. As discussed, deflate and base64 encoding will take care of most problems.

File: pkg/provider/xml/xml.go (hunk 1 above)
Comment (stebenz) [reply pkg/provider/xml/xml.go]: True, changed it to the suggestion with defer.

File: pkg/provider/response.go (hunk 2 above)
Comment (stebenz) [reply pkg/provider/response.go]: Renamed and refactored the usage of the expiration, also in the attributeQuery.

File: pkg/provider/response.go (hunk 3 above)
Comment (stebenz) [reply pkg/provider/response.go]: Rollbacked the changes here.

File: pkg/provider/xml/xml.go [hunk 7]
```diff
@@ -43,25 +44,19 @@ func Marshal(data interface{}) ([]byte, error) {
...
+	if err := w2.Close(); err != nil {
 		return nil, err
 	}
-
-	if err := w2.Flush(); err != nil {
+	if err := w1.Close(); err != nil {
 		return nil, err
 	}
-	return b.Bytes(), nil
+	return w.Bytes(), nil
```
[the remark this answers is not in the document]
Comment (stebenz) [reply pkg/provider/xml/xml.go]: Renamed the writer and decoder, added comment.

File: pkg/provider/provider.go [hunk 8]
```diff
@@ -217,12 +219,32 @@ var allowAllOrigins = func(_ string) bool {
 }
 
 // AuthCallbackURL builds the url for the redirect (with the requestID) after a successful login
-func AuthCallbackURL(p *Provider) func(context.Context, string) string {
+func (p *Provider) AuthCallbackURL() func(context.Context, string) string {
```
Comment (livio-a) [inline pkg/provider/provider.go:222]: just noticed now: any particular reason to switch from function to method?
(breaking change)

Comment (stebenz) [reply pkg/provider/provider.go:222]: To make it consistent to the OIDC usage, as I saw the difference

Comment (livio-a) [reply pkg/provider/provider.go:222]: but isn't it a function as well?
https://github.com/zitadel/oidc/blob/main/pkg/op/op.go#L150-L154

Comment (stebenz) [reply pkg/provider/provider.go:222]: This function is what I mean, do we have something else to change?
https://github.com/zitadel/oidc/blob/cf6ce69d79666cef2ac429607286c3aa63bcb677/pkg/op/server_legacy.go#L79

Comment (stebenz) [reply pkg/provider/provider.go:222]: https://github.com/zitadel/zitadel/blob/83bdaf43c35b7c4c5dfaf3e822a917e95d9f5fae/cmd/start/start.go#L532

**Review by livio-a — APPROVED** [review]

**Pull request merged**